core/src/main/java/org/apache/iceberg/encryption/EnvelopeEncryptionManager.java

```diff
@@ -0,0 +1,125 @@
+        "Cannot create EnvelopeEncryptionManager because table key encryption key ID is not specified");
+    this.kmsClient = Preconditions.checkNotNull(kmsClient,
+        "Cannot create EnvelopeEncryptionManager because KmsClient is null");
+    this.pushdown = pushdown;
+    this.dataKeyLength = dataKeyLength;
+  }
+
+  @Override
+  public EncryptedOutputFile encrypt(OutputFile rawOutput) {
+    EnvelopeMetadata metadata = generateEnvelopeMetadata(dataEncryptionConfig);
+
+    if (pushdown) {
+      // Pushdown to data formats with native encryption support (such as Parquet)
+      NativeFileCryptoParameters nativeEncryptParams = NativeFileCryptoParameters.create(metadata.dek())
+          .encryptionAlgorithm(metadata.algorithm().toString())
+          .build();
+
+      Preconditions.checkArgument(rawOutput instanceof NativelyEncryptedFile,
+          "Can't natively encrypt " + rawOutput.location() + " because the class " +
+              rawOutput.getClass() + " doesn't implement NativelyEncryptedFile interface");
+
+      ((NativelyEncryptedFile) rawOutput).setNativeCryptoParameters(nativeEncryptParams);
+
+      return new BaseEncryptedOutputFile(rawOutput, metadata);
+    } else {
+      throw new UnsupportedOperationException("EnvelopeEncryptionManager currently supports only pushdown " +
+          "data encryption / decryption");
+    }
+  }
+
+  @Override
+  public InputFile decrypt(EncryptedInputFile encrypted) {
+    EnvelopeMetadata metadata = EnvelopeMetadataParser.fromJson(encrypted.keyMetadata().buffer());
+    ByteBuffer fileDek = kmsClient.unwrapKey(metadata.wrappedDek(), metadata.kekId());
+
+    if (pushdown) {
+      // Pushdown to data formats with native decryption support (such as Parquet)
+      NativeFileCryptoParameters nativeDecryptParams = NativeFileCryptoParameters.create(fileDek).build();
+      InputFile rawInput = encrypted.encryptedInputFile();
+
+      Preconditions.checkArgument(rawInput instanceof NativelyEncryptedFile,
+          "Can't natively decrypt " + rawInput.location() + " because the class " +
+              rawInput.getClass() + " doesn't implement NativelyEncryptedFile interface");
+
+      ((NativelyEncryptedFile) rawInput).setNativeCryptoParameters(nativeDecryptParams);
+
+      return rawInput;
+    } else {
+      throw new UnsupportedOperationException("EnvelopeEncryptionManager currently supports only pushdown " +
+          "data encryption / decryption");
+    }
+  }
+
+  private EnvelopeMetadata generateEnvelopeMetadata(EnvelopeConfig config) {
+    String tableKekID = config.kekId();
+    byte[] fileDek;
+    String wrappedFileDEK;
+
+    if (kmsClient.supportsKeyGeneration()) {
+      KmsClient.KeyGenerationResult generatedDek = kmsClient.generateKey(tableKekID);
+      fileDek = generatedDek.key().array();
+      wrappedFileDEK = generatedDek.wrappedKey();
+    } else {
+      if (null == workerRNG) {
+        workerRNG = new SecureRandom();
```

Comment:
The Javadoc for this class should probably describe this behavior

When generating new keys for OutputFiles this class will first attempt have the KMS generate a new key, if the KMS does not support key generation a new key will be produced by pulling bytes from a <@link SecureRandom> on the JVM writing the file.

// I think this may be important to note for users running on virtualized systems which may not have good entropy, but I'm mostly guessing here.